File: LeetCode_Java/greedy/LemonadeChange.java
package greedy;

public class LemonadeChange {
    public boolean lemonadeChange(int[] bills) {
        int five = 0, ten = 0;

        for (int bill : bills) {
            if (bill == 5) {
                five++;
            } else if (bill == 10) {
                ten++;
                five--;
            } else {
                if (ten > 0) {
                    ten--;
                } else {
                    five -= 2;
                }
                five--;
            }

            if (five < 0) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
        LemonadeChange solution = new LemonadeChange();
        int[] bills1 = {5, 5, 5, 10, 20};
        System.out.println(solution.lemonadeChange(bills1));// Output: true

        int[] bills2 = {5, 5, 10, 20, 5, 5, 5, 5, 5, 5, 5, 5, 5, 10, 5, 5, 20, 5, 20, 5};
        System.out.println(solution.lemonadeChange(bills2));// Output: true

        int[] bills3 = {5, 5, 5, 10, 5, 5, 10, 20, 20, 20};
        System.out.println(solution.lemonadeChange(bills3));// Output: false
    }
}

File: LeetCode_Python/greedy/LemonadeChange.py
class LemonadeChange:
    def lemonadeChange(self, bills: list[int]) -> bool:
        five = ten = 0

        for bill in bills:
            if bill == 5:
                five += 1
            elif bill == 10:
                if five == 0:
                    return False
                five -= 1
                ten += 1
            else:
                if ten > 0 and five > 0:
                    ten -= 1
                    five -= 1
                elif five >= 3:
                    five -= 3
                else:
                    return False

        return True


# Time Complexity: O(n), where n is the number of bills
# Space Complexity: O(1), no extra space is used

if __name__ == "__main__":
    solution = LemonadeChange()

    print("Expected Output: True")
    print("Actual Output: ", solution.lemonadeChange([5, 5, 5, 10, 20]))

    print("Expected Output: True")
    print("Actual Output: ", solution.lemonadeChange([5, 5, 10]))

    print("Expected Output: False")
    print("Actual Output: ", solution.lemonadeChange([10, 10]))
